### PR TITLE
API-175: add integration tests for Product API

### PR DESCRIPTION
**Pull request by LaurentPetard**

### CODE REVIEW

File: tests/Api/Product/ListProductApiIntegration.php [hunk 1]
```diff
@@ -0,0 +1,1398 @@
...
+        $this->assertNull($firstPage->getCount());
+        $this->assertNull($firstPage->getPreviousLink());
+        $this->assertNull($firstPage->getPreviousPage());
+        $this->assertFalse($firstPage->hasPreviousPage());
+        $this->assertTrue($firstPage->hasNextPage());
+        $this->assertSame($baseUri . '/api/rest/v1/products?page=2&with_count=false&pagination_type=page&limit=5', $firstPage->getNextLink());
+
+        $products = $firstPage->getItems();
+        $this->assertCount(5 ,$products);
+        for ($i = 0; $i < 5; $i++) {
```
Comment (ahocquard) [inline tests/Api/Product/ListProductApiIntegration.php]: I would put it in a protected function `assertListResponse`.
```
$products = $firstPage->getItems();
$expectedProducts = array_slice($this->getExpectedProducts(), 0, 5);
$this->asssertListResponse($expectedProducts, $products);
```

And you can put the sanitize in the function.
It avoids `assertCount` as well.

File: tests/Api/Product/ListProductApiIntegration.php [hunk 2]
```diff
@@ -0,0 +1,1398 @@
...
+        $api = $this->createClient()->getProductApi();
+        $expectedProducts = $this->getExpectedProducts();
+        $baseUri = $this->getConfiguration()['api']['baseUri'];
+
+        $firstPage = $api->listPerPage(2, false, ['foo' => 'bar']);
+
+        $this->assertInstanceOf(PageInterface::class, $firstPage);
+        $this->assertSame($baseUri . '/api/rest/v1/products?page=2&with_count=false&pagination_type=page&limit=2&foo=bar', $firstPage->getNextLink());
+
+        $products = $firstPage->getItems();
```
Comment (ahocquard) [inline tests/Api/Product/ListProductApiIntegration.php]: same: `assertListResponse`

File: tests/Api/Product/ListProductApiIntegration.php [hunk 3]
```diff
@@ -0,0 +1,1398 @@
...
+    public function testAll()
+    {
+        $api = $this->createClient()->getProductApi();
+        $products = $api->all();
+        $expectedProducts = $this->getExpectedProducts();
+
+        $this->assertInstanceOf(ResourceCursorInterface::class, $products);
+
+        $productsCount = 0;
+        foreach ($products as $key => $product) {
```
Comment (ahocquard) [inline tests/Api/Product/ListProductApiIntegration.php]: same (and in the whole class).

File: tests/Api/Product/UpsertListProductApiIntegration.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                ]
+            ]
+        ]);
+
+        $this->assertInstanceOf('\Iterator', $response);
+
+        $responseLines = iterator_to_array($response);
+        $this->assertCount(2, $responseLines);
+
+        $this->assertSame([
```
Comment (ahocquard) [inline tests/Api/Product/UpsertListProductApiIntegration.php]: could be done in one assertSame without assertCount

File: tests/Api/Product/ListProductApiIntegration.php [hunk 5]
```diff
@@ -0,0 +1,1398 @@
...
+            $productsCount++;
+        }
+
+        $this->assertSame(10, $productsCount);
+    }
+
+    public function testSearchOnProductProperties()
+    {
+        $api = $this->createClient()->getProductApi();
+        $products = $api->listPerPage(10, true, [
```
Comment (skeleton) [inline tests/Api/Product/ListProductApiIntegration.php]: add a test to build the search with the search builder please

File: tests/Api/Product/ListProductApiIntegration.php [hunk 6]
```diff
@@ -0,0 +1,1398 @@
...
+
+        $this->assertSameContent($expectedProduct, $actualProduct);
+    }
+
+    public function testAllWithSelectedLocales()
+    {
+        $baseUri = $this->getConfiguration()['api']['baseUri'];
+        $api = $this->createClient()->getProductApi();
+        $products = $api->all(10, [
+            'locales' => 'fr_FR',
```
Comment (skeleton) [inline tests/Api/Product/ListProductApiIntegration.php]: it works ? locales should be an array, no ?

Comment (LaurentPetard) [inline tests/Api/Product/ListProductApiIntegration.php]: nope, separated by a comma if there are many. https://api.akeneo.com/documentation/filter.html#via-locale

File: tests/Api/Product/UpsertListProductApiIntegration.php (hunk 4 above)
Comment (LaurentPetard) [inline tests/Api/Product/UpsertListProductApiIntegration.php]: withe the rework of sanitizeProductData it's even simpler than that

File: tests/Api/Product/ListProductApiIntegration.php (hunk 6 above)
Comment (ahocquard) [inline tests/Api/Product/ListProductApiIntegration.php]: It should be modified when we will use fluid methods.

**Pull request merged**